src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceInputFormat.java

```diff
@@ -135,4 +132,15 @@ private BulkConnection getBulkConnection(Configuration conf) {
     }
   }
 
+  private RecordReader<Schema, Map<String, ?>> getDelegateRecordReader(String query, Schema schema) {
+    if (SalesforceQueryParser.isRestrictedQuery(query)) {
+      LOG.warn("Restricted SOQL query detected switching to SOAP reader. Performance can be significantly impacted");
+      return new SalesforceSoapRecordReader(schema, query, new SoapRecordToMapTransformer());
+    }
+    if (SalesforceQueryUtil.isQueryUnderLengthLimit(query)) {
+      return new SalesforceBulkRecordReader(schema);
+    }
+    LOG.warn("Wide SOQL query detected partially switching to SOAP reader. Performance may be impacted");
```

Comment:
The SOQL query is a wide query. An additional SOAP request will be performed for each record.